libraries/AP_Avoidance/AP_Avoidance.cpp

```diff
@@ -0,0 +1,598 @@
+#include "AP_Avoidance.h"
+
+extern const AP_HAL::HAL& hal;
+
+#include <limits>
+#include "distance.h"
+#include <GCS_MAVLink/GCS.h>
+
+AP_Avoidance *AP_Avoidance::_instance;
+
+#include "AvoidanceHandler.h"
+
+#define AVOIDANCE_DEBUGGING 0
+
+#if AVOIDANCE_DEBUGGING
+#include <stdio.h>
+#define debug(fmt, args ...)  do {::fprintf(stderr,"%s:%d: " fmt "\n", __FUNCTION__, __LINE__, ## args); } while(0)
+#else
+#define debug(fmt, args ...)
+#endif
+
+// table of user settable parameters
+const AP_Param::GroupInfo AP_Avoidance::var_info[] = {
+
+    // @Param: ENABLE
+    // @DisplayName: Enable AVOIDANCE
+    // @Description: Enable Avoidance
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO("ENABLE",     0, AP_Avoidance, _enabled,    0),
+
+    // @Param: F_ACTION
+    // @DisplayName: Collision Avoidance Behavior
+    // @Description: Specifies aircraft behaviour when a collision is imminent
+    // The following values should come from the mavlink COLLISION_ACTION enum
+    // @Values: 0:None,1:Report,4:TCAS,5:Perpendicular,6:RTL,7:Hover
+    // @User: Advanced
+    AP_GROUPINFO("F_ACTION",   1, AP_Avoidance, _fail_action, MAV_COLLISION_ACTION_REPORT),
+
+    // @Param: W_BEHAVIOR
+    // @DisplayName: Collision Avoidance Behavior - Warn
+    // @Description: Specifies aircraft behaviour when a collision may occur
+    // The following values should come from the mavlink COLLISION_ACTION enum
+    // @Values: 0:None,1:Report,4:TCAS,5:Perpendicular,6:RTL,7:Hover
+    // @User: Advanced
+    AP_GROUPINFO("W_ACTION",   2, AP_Avoidance, _warn_action, MAV_COLLISION_ACTION_REPORT),
+
+    // @Param: RCVRY_W
+    // @DisplayName: Recovery behaviour after a warn event
+    // @Description: Determines what the aircraft will do after a warn event (which has not progressed to a fail event) has been resolved
+    // @Values: 0:Resume previous flight mode,1:Mode Loiter,2:Continue collision avoidance action
+    // @User: Advanced
+    AP_GROUPINFO("W_RCVRY",   3, AP_Avoidance, _warn_recovery, AVOIDANCE_RECOVERY_W_RESUME),
+
+    // @Param: RCVRY_F
+    // @DisplayName: Recovery behaviour after a fail event
+    // @Description: Determines what the aircraft will do after a fail event is resolved
+    // @Values: 0:Continue collision avoidance ACTION_F,1:Move to collision avoidance W_ACTION
+    // @User: Advanced
+    AP_GROUPINFO("F_RCVRY",   4, AP_Avoidance, _fail_recovery, AVOIDANCE_RECOVERY_F_MOVE_TO_WARN),
+
+    // @Param: OBS_MAX
+    // @DisplayName: Maximum number of obstacles to track
+    // @Description: Maximum number of obstacles to track
+    // @User: Advanced
+    AP_GROUPINFO("OBS_MAX",   5, AP_Avoidance, _samples_max, 20),
```

Comment:
samples is a weird name to use, right in the desc you say "obstacles" which is a much better name